### PR TITLE
Replace B+tree with ordered dictionary

### PR DESCRIPTION
**Pull request by jxsvoboda**

### CODE REVIEW

File: kernel/generic/src/mm/as.c [hunk 1]
```diff
@@ -1866,293 +1853,164 @@ size_t as_area_get_size(uintptr_t base)
...
-	panic("Inconsistency detected while adding %zu pages of used "
-	    "space at %p.", count, (void *) page);
+static void used_space_remove_ival(used_space_ival_t *ival)
+{
+	odict_remove(&ival->lused_space);
+	slab_free(used_space_ival_cache, ival);
+}
 
-success:
-	area->resident += count;
```
Comment (jermar) [inline kernel/generic/src/mm/as.c]: I think the new code does not track area->resident anymore. This probably breaks statistics.

File: kernel/generic/src/mm/backend_elf.c [hunk 2]
```diff
@@ -163,72 +163,65 @@ void elf_share(as_area_t *area)
...
 			/*
-			 * Skip read-only areas of used space that are backed
-			 * by the ELF image.
+			 * Skip read-only pages that are backed by the
+			 * ELF image.
 			 */
 			if (!(area->flags & AS_AREA_WRITE))
 				if (base >= entry->p_vaddr &&
-				    base + P2SZ(count) <= start_anon)
+				    base + P2SZ(i + 1) <= start_anon)
```
Comment (jermar) [inline kernel/generic/src/mm/backend_elf.c]: Could this possibly overflow?

File: kernel/generic/src/mm/as.c [hunk 3]
```diff
@@ -2163,175 +2021,65 @@ bool used_space_insert(as_area_t *area, uintptr_t page, size_t count)
...
+		odict_insert(&ival->lused_space, &area->used_space.ivals,
+		    NULL);
 	}
 
-error:
-	panic("Inconsistency detected while removing %zu pages of used "
-	    "space from %p.", count, (void *) page);
-
-success:
-	area->resident -= count;
```
Comment (jermar) [inline kernel/generic/src/mm/as.c]: Again, this seems to get lost in the new version.

File: kernel/generic/src/mm/backend_elf.c (hunk 2 above)
Comment (jxsvoboda) [reply kernel/generic/src/mm/backend_elf.c]: I don't know but I am not changing the logic in this piece of code. Just indentation and j -> i. The diff is mismatched.

File: kernel/generic/include/mm/as.h [hunk 4]
```diff
@@ -157,6 +156,27 @@ typedef struct as_pagemap {
...
+	struct used_space *used_space;
+	/** Link to @c used_space->ivals */
+	odlink_t lused_space;
+	/** First page address */
+	uintptr_t page;
+	/** Count of pages */
+	size_t count;
+} used_space_ival_t;
+
+/** Map of used space in an adress space area */
```
[the remark this answers is not in the document]
Comment (jxsvoboda) [reply kernel/generic/include/mm/as.h]: OK

File: kernel/generic/src/mm/as.c (hunk 1 above)
Comment (jxsvoboda) [reply kernel/generic/src/mm/as.c]: OK, I think I need to move the used page count into used_space_t.

File: kernel/generic/src/mm/as.c (hunk 3 above)
Comment (jxsvoboda) [reply kernel/generic/src/mm/as.c]: OK.

**Comment by jxsvoboda** [comment]
I made fixes according to your comments. Number of resident pages is now tracked in used_space_t.pages. Fixed the typos and added missing documentation to functions used_space_remove_ival() and used_space_shorten_ival().

File: kernel/generic/src/mm/as.c [hunk 5]
```diff
@@ -1398,16 +1390,14 @@ errno_t as_area_change_flags(as_t *as, unsigned int flags, uintptr_t address)
...
 
-	list_foreach(area->used_space.leaf_list, leaf_link, btree_node_t,
-	    node) {
-		btree_key_t i;
-
-		for (i = 0; i < node->keys; i++)
-			used_pages += (size_t) node->value[i];
+	used_space_ival_t *ival = used_space_first(&area->used_space);
+	while (ival != NULL) {
+		used_pages += ival->count;
```
Comment (jermar) [inline kernel/generic/src/mm/as.c]: Is the computed value different from the one stored in area->used_space.pages?

Comment (jxsvoboda) [reply kernel/generic/src/mm/as.c]: No, it's the same. I will change it to simply use  used_space.pages.

**Pull request merged**